File: taojieTan/assignments/cyclicSort/lc268/lc268.py
# Problem Statement #

# We are given an array containing ‘n’ distinct numbers taken from the range 0 to ‘n’. Since the array has only ‘n’ numbers out of the total ‘n+1’ numbers, find the missing number.

# input: an unsorted list with integers ranging from 0 to n and length of n
# output: an integer
# edge case: empty list => return -1

'''
 We can use a while loop swap each integer into its correct position (integer 1 will be on index 0, 2 will be on index 1, and so on)
 and we need to skip the integer zero. 
 After the while loop, the position where integer zero sits will be the missing integer (index + 1)
 We can use the for loop to traverse through the list to check the zero integer.
'''


def find_missing_number(nums):
  # TODO: Write your code here
  index = 0 

  while index < len(nums):
    if nums[index] != index + 1 and nums[index] != 0:
      correct_pos = nums[index] - 1
      nums[correct_pos], nums[index] = nums[index], nums[correct_pos]
    else:
      index += 1

  for i in range(len(nums)):
    if nums[i] == 0:
      return i+1

  return -1

def main():
  print(find_missing_number([4, 0, 3, 1])) 
  print(find_missing_number([8, 3, 5, 2, 4, 6, 0, 1]))
  print(find_missing_number([0, 1]))

main()




# Solution
# -----
# def find_missing_number(nums):
#   i, n = 0, len(nums)
#   while i < n:
#     j = nums[i]
#     if nums[i] < n and nums[i] != nums[j]:
#       nums[i], nums[j] = nums[j], nums[i]  # swap
#     else:
#       i += 1

#   # find the first number missing from its index, that will be our required number
#   for i in range(n):
#     if nums[i] != i:
#       return i

#   return n
  
# -----

# Time complexity #
# The time complexity of the above algorithm is O(n). In the while loop, although we are not incrementing the index i when swapping the numbers, this will result in more than n iterations of the loop, but in the worst-case scenario, the while loop will swap a total of n-1 numbers and once a number is at its correct index, we will move on to the next number by incrementing i. In the end, we iterate the input array again to find the first number missing from its index, so overall, our algorithm will take O(n) + O(n-1) + O(n) which is asymptotically equivalent to O(n).

# Space complexity #
# The algorithm runs in constant space O(1).

File: tiantianXia/assignments/cyclicSort/lc268/lc268.js
// Problem Statement #

// We are given an array containing ‘n’ distinct numbers taken from the range 0 to ‘n’. Since the array has only ‘n’ numbers out of the total ‘n+1’ numbers, find the missing number.

/**
 * inputs: 
 * @param {*} nums : an array that containing distinct numbers and unsorted. the range could be from 0 - n and may not continues
 * 
 * output
 * @returns missNum: look for the missing number
 * 
 * step 1: loop all the elements
 * step 2: compare current pointer with current poiner to determine if the number is at the right postion if not swap
 * step 3: after swap if the current pointer is not equal to the value return the value
 * 
 * variables: currentPrt: a potiner that return the current value
 * 
 * a while loop that traverse all elements and check if the currentPrt is less than the length of nums while true
 *    define a swapPrt to be the value of the current pointer
 *    compare if the value current pointer is not equal to the value of the current pointer
 *        do swap with value in the swapPrt and the current prt vale
 *    else
 *        increment the currentprt
 * end loop
 * 
 * a for loop to traverse all the elements to find out the missing number
 *  compare if the current pointer is equal to the vaule of the current pointer if not equal
 *      reutn the current pointer meanning that we find the missing number
 * end loop
 * 
 * edge case: if we don't find the number return -1
 *            if num = [] => -1
 */
const find_missing_number = function(nums) {

  if(!nums.length) return -1;
  // TODO: Write your code here

  let currentPrt = 0;

  while(currentPrt < nums.length){

    const swapPrt = nums[currentPrt];

    if(nums[currentPrt] !== nums[swapPrt]){
      [nums[currentPrt], nums[swapPrt]] = [nums[swapPrt], nums[currentPrt]];
    }else{
      currentPrt ++;
    }
  }

  for(let i = 0; i < nums.length; i++){
    if(i !== nums[i]) return i;
  }

  return -1;
};

console.log(find_missing_number([4, 0, 3, 1])); //2
console.log(find_missing_number([8, 3, 5, 2, 4, 6, 0, 1])); // 7
console.log(find_missing_number([]));

// Solution
// -----
// function find_missing_number(nums) {
//   let i = 0;
//   const n = nums.length;
//   while (i < n) {
//     j = nums[i];
//     if (nums[i] < n && nums[i] !== nums[j]) {
//       [nums[i], nums[j]] = [nums[j], nums[i]]; // swap
//     } else {
//       i += 1;
//     }
//   }

//   // find the first number missing from its index, that will be our required number
//   for (i = 0; i < n; i++) {
//     if (nums[i] !== i) {
//       return i;
//     }
//   }

//   return n;
// }

// -----

// Time complexity #
// The time complexity of the above algorithm is O(n). In the while loop, although we are not incrementing the index i when swapping the numbers, this will result in more than n iterations of the loop, but in the worst-case scenario, the while loop will swap a total of n-1 numbers and once a number is at its correct index, we will move on to the next number by incrementing i. In the end, we iterate the input array again to find the first number missing from its index, so overall, our algorithm will take O(n) + O(n-1) + O(n) which is asymptotically equivalent to O(n).

// Space complexity #
// The algorithm runs in constant space O(1).
